### PR TITLE
`RawTask` refactoring

### PR DESCRIPTION
**Pull request by oliver-giersch**
Looking through the code of this crate, I've found the the `raw` module to be quite obtuse and unnecessarily so.

Basically, what this PR does is refactor the `raw` module and its explicit and implicit dependencies in the rest of the crate for greater clarity, and in the process achieves a net reduction of approximately 100 lines of code by removing most of the manual memory layout fiddling.

At its core, `RawTask` is now no longer a collection of pointers into a manually allocated and otherwise undifferentiated blob of memory on the heap.
Instead, it is now a `#[repr(C)]` struct describing the actual memory layout of a task.
I assume, this was originally done because `union`s with non-Copy fields are unstable, but I believe the small overhead of using a (tagged) `enum` for the Future/Output is outweighed by the clearer and more concise code.
Either way, using `async-task`'s own benchmark suite, I found no performance difference between either variant.
Once non-Copy `unions` become stable, the current code can be easily refactored to use a `union` instead of an `enum`.

### CODE REVIEW

**Comment by skade** [comment]
Hi @oliver-giersch! This looks interesting, but actually fails one of the memleak tests!

https://github.com/async-rs/async-task/pull/19/checks?check_run_id=519509441

**Comment by oliver-giersch** [comment]
> 
> 
> Hi @oliver-giersch! This looks interesting, but actually fails one of the memleak tests!
> 
> https://github.com/async-rs/async-task/pull/19/checks?check_run_id=519509441

Oh, thanks for the heads up, I hadn't noticed and will look into it ASAP.

**Comment by skade** [comment]
@oliver-giersch No problem, that's what we have the CI for :).

**Comment by skade** [comment]
There's an interesting spurious failure in the test suite here: https://travis-ci.com/github/async-rs/async-task/jobs/316888350

This might hint to a bug...

**Comment by oliver-giersch** [comment]
Locally, all tests passed for using (using lsan instead of valgrind). I'll check what might have caused this CI failure, or do you actually think it is a spurious failure?

**Comment by skade** [comment]
It's spurious, but shouldn't happen, which makes me uneasy.

If you look at the build matrix, you'll notice that we run on windows tests twice in the same way (that's a current CI bug), but only one fails.

https://travis-ci.com/github/async-rs/async-task/builds/159307360

**Comment by oliver-giersch** [comment]
Apologies for not following through on this, I had tried seeing through the failing test in question and had failed back then and now again recently. It's very hard to understand what the test even tries to test (with all the macro invocations involved). Then there's a bunch of dependencies used by the tests themselves, some of which are pretty complicated concurrency related utilities, all of which could potentially also cause the failure in question. Also, the test only failed on Windows when using Valgrind, so it could perhaps even be an issue with Valgrind. 

To summarize, I don't feel like I have the capacities to debug this issue in the depth that would be required.

**Comment by skade** [comment]
@oliver-giersch I'll close it then. async-task mainline has actually moved to @stjepang.
